### ISSUE TITLE
Google-style: New Check TextBlockGoogleStyleFormattingCheck

### ISSUE THREAD

**Issue by mohitsatr** [checkstyle/checkstyle#17223]
From #17159 

`TextBlockGoogleStyleFormattingCheck` should validate if the Opening and closing quotes of java text-blocks are on new lines. Google-style demands opening and closing (""") should always be on new lines.

https://google.github.io/styleguide/javaguide.html#s4.8.9-text-blocks
> The opening """ of a text block is always on a new line. That line may either follow the same indentation rules as other constructs, or it may have no indentation at all (so it starts at the left margin). The closing """ is on a new line with the same indentation as the opening """, and may be followed on the same line by further code

Example Config:
```xml
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="TextBlockGoogleStyleFormattingCheck"/>
    </module>
</module>
```

TextBlocks.java Example:
```java
$ cat TextBlocks.java 
public class TextBlocks {

  public static String textFun(String[] args) {

    // violation below, 'Opening quotes (""") of text-block must be on the new line.'
    final String simpleScript = """ 
        s
        """;  // OK

    final String simpleScript1 = 
        """  // OK
        this is simple test;
        """;  // OK

    final String simpleScript2 = 
        """  // OK
        this is sample text """; 
        // violation above, 'Opening quotes (""") of text-block must be on the new line.'
    
    String s = 
            """
              Hello there
            """ + getName();  // OK 

    // violation below, 'Opening quotes (""") of text-block must be on the new line.'  
    return """
        this is sample text 
        """;  // OK
  }
}

```

Expected Output:
```
$ java -jar testing/checkstyle-10.25.0-all.jar -c config.xml TextBlocks.java 
Starting audit...
[WARN] TextBlocks.java:5: Opening quotes (""") of text-block must be on the new line. [TextBlockGoogleStyleFormattingCheck]
[WARN] TextBlocks.java:17: Closing quotes (""") of text-block must be on the new line. [TextBlockGoogleStyleFormattingCheck]
[WARN] TextBlocks.java:28: Opening quotes (""") of text-block must be on the new line. [TextBlockGoogleStyleFormattingCheck]
Audit done.
```
Some other examples showing how the check should work

CASE 1: Placement of opening quotes:

```java
  public static String textFun() { 

    // violation below, 'Opening quotes (""") of text-block must be on the new line.'
    final String simpleScript = """ 
        s
        """;

    // OK
    final String simpleScript1 = 
        """  
        this is simple test;
        """;

    // violation below, 'Opening quotes (""") of text-block must be on the new line.'
    getData(""" 
        Hello,
        This is a multi-line message.
    """);
    
    // violation below, 'Opening quotes (""") of text-block must be on the new line.'
    return """
        this is sample text 
        """;
  } 
```

CASE 2: Placement of Closing quotes:

```java
  public String textFun2() {

    final String simpleScript2 = 
      """
        this is sample text """;  
        // violation above, 'Closing quotes (""") of text-block must be on the new line.'

      getData(
          """
          Hello,
          This is a multi-line message."""
          // violation above, 'Closing quotes (""") of text-block must be on the new line.'
      );

      return 
          """
            THE MULTI-LINE MESSAGE """; // violation, 'Closing quotes (""") of text-block must be on the new line.'
  }
```

CASE 3: Closing quotes followed by some expression

```java
  public String textFun3() {

    String s = 
        """
          Hello there
        """ + getName();  // OK

    getData(
      """
        hello there1
      """, 0);  // OK

    return s + 
        """
        very good
        """.charAt(0) + getName();  // OK
  } 
```

CASE 4 : Indentation
Checking indentation of text-block is the second function of this Check.

> That line may either follow the same indentation rules as other constructs, or it may have no indentation at all (so it starts at the left margin).
 
Opening quotes can have indentation of 0 or indentation value of parent + `lineWrappingIndentation`. 
Value of `lineWrappingIndentation` is 4 in style-guide.
For closing quotes, we just need to make sure they are vertically aligned with opening quotes.

```java
  public void textFun4() {
    
    // OK 
    String e1 =
        """
      content of the block. e1
        """;

    // According to rule following is OK, but code does not look clean. 
    String e2 = 
"""      
         content of the block e2
""";

    String e3 = 
        """
        content of the block of e3
            """;
    // violation above, 'Text-block quotes are not vertically aligned. '

    // OK
    getData(
      """
        Indentation of Text-block 
      """,
      5
    );
  }
```

Behaviour of Indentation Check on Java textblocks: Indentation Check only gives violation if actual indentation of opening or closing quotes is less can expected indentation

CASE 4a: indenation is more than expected

```java
/** somejavadoc. */
public class TextBlockIndentation {

  /** somejavadoc. */
  public void textIndentation1() {
    String e1 =
        """
      content of the block. e1
        """;

    String e2 = 
        """
        content of the block of e3 
            """; // EXPECTED VIOLATION ON CLOSING """
    
    // EXPECTED VIOLATION ON OPENING """
    String e3 =
          """ 
      content of the block. e1
        """;
    
    // EXPECTED VIOLATION ON OPENING AND CLOSING """s
    getData(
            """
        Indentation of Text-block 
            """,
        5
    );   
  }
```

```
$ java -jar checkstyle-10.25.0-all.jar -c google_checks.xml TextBlockIndentation.java 
Starting audit...
Audit done.
```

Google Formatter:

```java
$ java -jar google-java-format-1.27.0-all-deps.jar TextBlockIndentation.java > FormattedCode.java && diff -Naru TextBlockIndentation.java FormattedCode.java 
--- TextBlockIndentation.java   2025-06-27 18:35:13.709744600 +0530
+++ FormattedCode.java  2025-06-27 18:54:42.372702500 +0530
@@ -6,27 +6,26 @@
     // OK
     String e1 =
         """
-      content of the block. e1
+        content of the block. e1
         """;
 
-    String e2 = 
+    String e2 =
         """
-        content of the block of e3 
-            """; // EXPECTED VIOLATION
-    
+        content of the block of e3
+        """; // EXPECTED VIOLATION
+
     // EXPECTED VIOLATION
     String e3 =
-          """ 
-      content of the block. e1
+        """
+        content of the block. e1
         """;
-    
+
     // EXPECTED VIOLATION
     getData(
-            """
-        Indentation of Text-block 
-            """, // EXPECTED VIOLATION
-        5
-    );   
+        """
+        Indentation of Text-block
+        """, // EXPECTED VIOLATION
+        5);
   }
```
CASE 4b: Indentation is more than expected

In the first example, indentation of both the quotes is 0 which is valid according to the rule, but Indentation Check, as expected, gives violation on it.
```java
  /** somejavadoc. */
  public void textFuncIndenation2() {
    String e2 =
"""
content of the block e2
""";
    
    getData(  // OK
    """ 
        Indentation of Text-block 
    """,
        5
    );
  }
```

```
Starting audit...
[WARN] /mnt/5D92528E6B945467/test/testing/TextBlockIndentation.java:4:1: '"""' has incorrect indentation level 0, expected level should be 8. [Indentation]
[WARN] /mnt/5D92528E6B945467/test/testing/TextBlockIndentation.java:6:1: '"""' has incorrect indentation level 0, expected level should be 8. [Indentation]
[WARN] /mnt/5D92528E6B945467/test/testing/TextBlockIndentation.java:9:5: 'method call' child has incorrect indentation level 4, expected level should be 6. [Indentation]
[WARN] /mnt/5D92528E6B945467/test/testing/TextBlockIndentation.java:11:5: 'method call' child has incorrect indentation level 4, expected level should be 6. [Indentation]
Audit done.
```
Google Formatter:

```java
$ java -jar google-java-format-1.27.0-all-deps.jar TextBlockIndentation.java > FormattedCode.java && diff -Naru TextBlockIndentation.java FormattedCode.java 
--- TextBlockIndentation.java   2025-06-27 18:35:13.709744600 +0530
+++ FormattedCode.java  2025-06-27 18:59:26.774067900 +0530
   /** somejavadoc. */
@@ -36,16 +35,13 @@
 content of the block e2
 """;
 
-    
     getData(
-    """
-        Indentation of Text-block 
-    """,
-        5
-    );
+        """
+            Indentation of Text-block
+        """,
+        5);
   }
 }
```
FYI, Indentation Check ignores the content of text-block, and that's fine because style guide does not mention it either.

**Comment by Zopsss** [comment]
Also add examples of incorrect indentation

**Comment by mohitsatr** [comment]
@Zopsss are we also concerned about the indentation value ? I thought this check's job would be limited to checking the position of the quotes--whether they are on new line or not.

**Comment by Zopsss** [comment]
> @Zopsss are we also concerned about the indentation value ?

Yes

> I thought this check's job would be limited to checking the position of the quotes--whether they are on new line or not.

We're implementing this check specifically for google style so we should implement it in such a way that it would cover the whole rule properly. 

We can make a property to check for Indentation, if that property is enable only then we should check for Indentation. We can do this for enforcing `"""` to be on new line as well. It'll make the Check flexible so user can use it in any way he likes.

**Comment by Zopsss** [comment]
Please explain all the functionalities the Check should follow with examples in Issue description @mohitsatr 

This should be a good example of how to propose a new check: https://github.com/checkstyle/checkstyle/issues/14415

**Comment by romani** [comment]
this check is very Google specific, lets name it as `TextBlockGoogleStyleFormattingCheck`
to be very very clear that it will serve all just for Google. If somebody like it, they can reuse it, but no customization to Check is expected by design.

**Comment by Zopsss** [comment]
> TextBlockGoogleStyleFormattingCheck

We shouldn't add `GoogleStyle` in check name, no check is named like this and it'll be confusing for users too if they're not familiar with google style guide

**Comment by romani** [comment]
I know, but all this requirements from Google on formatting is easy to justify, it is just "I like this way in Google" .

We name Checks without "GoogleStyle" only if make rule general and reusable by others. We did a lot for this, but still some of them are still mostly for Google and when google change style and we need change Check, we do excuse "we did Check for google style, so we just changing behavior even thee might be some people who do not like new behavior."

This Style pharagraph has a lot "and ....  and ..... and .....". 

To avoid GoogleStyle we can go deeper analysis and split requirements in multiple Checks and think deeper. But we can do this later on at any point of time if user experince interest. and just have Check with "GoogleStyle" in name that does all as style guide wants.

**Comment by mohitsatr** [comment]
@Zopsss @romani please review the issue description and let me know what feels missing.

I still have doubts regarding this Check handling indentation. we need to think how to avoid conflict with existing Indentation check. since this check's primarily going to be used for google-style, It may not be good idea to modify `IndentationCheck` to ignore Text-blocks.

**Comment by romani** [comment]
>`<property name="indentationAware" value="`

Vs 

>Opening quotes can have indentation of 0 or indentation value of parent + lineWrappingIndentation.

Did you miss renaming any of this phrase?

**Comment by romani** [comment]
>That line may either follow the same indentation rules as other constructs, or it may have no indentation at all (so it starts at the left margin).

As we made Check to Google style specific, we do not need properties at all.we just do validation as Google style declares.

**Comment by mohitsatr** [comment]
> As we made Check to Google style specific, we do not need properties at all.we just do validation as Google style declares.

I have updated the description. please check

**Comment by Zopsss** [comment]
Description looks good to me expect the Check name. I understand what @romani was trying to explain at https://github.com/checkstyle/checkstyle/issues/17223#issuecomment-2994218985, he is correct that this check is specifically for Google. But I don't like this name. Other than this, I'm fine with the current working of Check.

**Comment by Zopsss** [comment]
> we need to think how to avoid conflict with existing Indentation check. since this check's primarily going to be used for google-style, It may not be good idea to modify IndentationCheck to ignore Text-blocks.

Can you show how the Indentation check is behaving with Text blocks? So we can implement this new check accordingly

**Comment by romani** [comment]
We can try to use xpath suppression to avoid some specific violations.

Cam you share behavior on Google style input in issue description. To make yo clear what we have to deal with. Please use formatter tool also.

**Comment by mohitsatr** [comment]
@Zopsss I updated the description to show Indentation behaviour. please review

**Comment by romani** [comment]
I generally fine with described behavior in description.

I see no suppression in Google checks https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml

Looks like Indentation Check just simply skip validation at all on open `"""`. But we would like it to be , except for cases with open `"""` is starting line with 0 indentation.
 Sounds like we would rather keep Indentation to skip open `"""` , just do this validation in new Check.

If open `"""` even being required to be validated in regular way, we will need to add xpath suppression to Google config to let very special Check handling it.

Google formatter does a bit more strict formatting in single way. It is ok, it is his business. We just need to give no violations on auto formatted code.

**Comment by romani** [comment]
@mohitsatr , can you look at Indentation to confirm that  there will be no violations from it for now ever.

**Comment by mohitsatr** [comment]
@romani  Indentation Check is producing violation only if `"""`'s indentation is less than expected. so, for example, if the expected indentation is 12, any level less than 12 will trigger violations from Indentation Check.
 
if the `"""`'s indentation is more than expected, there will be no violations. that is because in google-config `forceStrictCondition` property of Indentation Check is off by default and it is good for us. 

so it is not right to say there will be no violations from Indentation Check. I might have to suppress some violations from it

**Comment by romani** [comment]
Ok, let's start doing Check and inputs for it, and we will do a bit more efforts in iT tests inputs..

I am ok with design. I think we covered all big items, small items of implementation wio be discussed as we start coding.

@Zopsss , please do final review.

**Comment by Zopsss** [comment]
The following case is from the first example:

```java
// violation below, 'Opening quotes (""") of text-block must be on the new line.'
    final String simpleScript = """ 
        s
        """;  // OK
```

@mohitsatr , fyi this should give 2 violations:

1. Opening `"""` should be on new line 
2. `"""` are not vertically aligned. 


Many such cases didn't mentioned about not vertically aligned `"""` violation but that's fine because we already have a separate example for it. I just pointed out that there should be 2 violations instead of 1.

**Comment by Zopsss** [comment]
Let's start implementing this Check

**Comment by romani** [comment]
fix is merged.

@mohitsatr , @Zopsss  thanks a lot for your dedication  and time to make it happen.

**Comment by mohitsatr** [comment]
many thanks to you @romani
